test/regression/riscv/instructions/arithmetic.cc

```diff
@@ -0,0 +1,365 @@
+      srai t6, t4, 1
+      addi t4, t4, 8
+      sra t1, t4, t3
+      srai t2, t4, 1
+  )");
+  EXPECT_EQ(getGeneralRegister<uint64_t>(30), -1);
+  EXPECT_EQ(getGeneralRegister<uint64_t>(31), -2);
+  EXPECT_EQ(getGeneralRegister<uint64_t>(6), 1);
+  EXPECT_EQ(getGeneralRegister<uint64_t>(7), 2);
+}
+
+TEST_P(InstArithmetic, sraw) {
+  RUN_RISCV(R"(
+      addi t3, t3, 2
+      addi t4, t4, 1
+      slli t5, t4, 31
+      li t2, 31
+      sraw t5, t5, t2
+
+      li t2, 30
+      slli t6, t4, 30
+      sraw t6, t6, t2
+  )");
+  EXPECT_EQ(getGeneralRegister<uint64_t>(30), -1);
+  EXPECT_EQ(getGeneralRegister<uint64_t>(31), 1);
+
+  RUN_RISCV(R"(
+      addi t3, t3, 2
+      addi t4, t4, 1
+      slli t5, t4, 31
+      sraiw t5, t5, 31
+      slli t6, t4, 30
+      sraiw t6, t6, 30
+  )");
+  EXPECT_EQ(getGeneralRegister<uint64_t>(30), -1);
+  EXPECT_EQ(getGeneralRegister<uint64_t>(31), 1);
+}
+
+TEST_P(InstArithmetic, add) {
+  RUN_RISCV(R"(
+      addi t3, t3, 3
+      addi t4, t4, 6
+      add t5, t3, t4
+      addi zero, t4, 16
+  )");
+  EXPECT_EQ(getGeneralRegister<uint64_t>(28), 3u);
+  EXPECT_EQ(getGeneralRegister<uint64_t>(29), 6u);
+  EXPECT_EQ(getGeneralRegister<uint64_t>(30), 9u);
+  EXPECT_EQ(getGeneralRegister<uint64_t>(0), 0);
+}
+
+TEST_P(InstArithmetic, addw) {
+  RUN_RISCV(R"(
+    addi t2, t2, -7
+    addi t3, t3, 3
+    addi t4, t4, 6
+    addw t5, t3, t4
+    addw t6, t2, t3
+  )");
+  EXPECT_EQ(getGeneralRegister<uint64_t>(28), 3u);
+  EXPECT_EQ(getGeneralRegister<uint64_t>(29), 6u);
+  EXPECT_EQ(getGeneralRegister<uint64_t>(30), 9u);
+  EXPECT_EQ(getGeneralRegister<uint64_t>(31), -4);
+}
+
+TEST_P(InstArithmetic, addiw) {
+  RUN_RISCV(R"(
+    addi t3, t3, 91
+    slli t3, t3, 28
+    addi t4, t4, -5
+    addiw t5, t3, -5
+    addiw t6, t2, -5
+  )");
+  EXPECT_EQ(getGeneralRegister<uint64_t>(28), 24427626496);
+  EXPECT_EQ(getGeneralRegister<int64_t>(29), -5);
+  EXPECT_EQ(getGeneralRegister<int32_t>(30), -1342177285);
+  EXPECT_EQ(getGeneralRegister<int64_t>(31), -5);
+}
+
+TEST_P(InstArithmetic, sub) {
+  RUN_RISCV(R"(
+    addi t3, t3, 3
+    addi t4, t4, 6
+    sub t5, t3, t4
+    sub t6, t4, t3
+  )");
+  EXPECT_EQ(getGeneralRegister<uint64_t>(30), -3);
+  EXPECT_EQ(getGeneralRegister<uint64_t>(31), 3);
+}
+
+TEST_P(InstArithmetic, subw) {
+  RUN_RISCV(R"(
+    addi t3, t3, 3
+    addi t4, t4, 6
+    subw t5, t3, t4
+    subw t6, t4, t3
+
+    li t3, -1
+    addi t4, t4, -8
+    subw t1, t3, t4
+
+  )");
+  EXPECT_EQ(getGeneralRegister<uint64_t>(30), -3);
+  EXPECT_EQ(getGeneralRegister<uint64_t>(31), 3);
+
+  EXPECT_EQ(getGeneralRegister<uint64_t>(28), 0xFFFFFFFFFFFFFFFF);
+  EXPECT_EQ(getGeneralRegister<uint64_t>(29), -2);
+  EXPECT_EQ(getGeneralRegister<uint64_t>(6), 0x0000000000000001);
+}
+
+TEST_P(InstArithmetic, lui) {
+  RUN_RISCV(R"(
+      lui t3, 4
+      lui t4, 0xFFFFC
+  )");
+  EXPECT_EQ(getGeneralRegister<uint64_t>(28), 4 << 12);
+  EXPECT_EQ(getGeneralRegister<uint64_t>(29), -4 << 12);
```

Comment:
Left shift of negative number is undefined behaviour.  You could try `-4ull << 12`.